When qualifying Clang types with a module, make sure we choose a visible module

martinboehme opened the pull request:
Clang types need special treatment because multiple Clang modules can contain the same type declarations from a textually included header, but not all of these modules may be visible.

Fixes #55477

The newly added test breaks without this fix.

martinboehme committed:
Repro for an invalid .swiftinterface generated when textually including a C
struct in different modules.

The newly added test fails.

martinboehme committed:
When qualifying Clang types with a module, make sure we choose a visible
module.

Clang types need special treatment because multiple Clang modules can
contain the same type declarations from a textually included header, but
not all of these modules may be visible.

This fixes

https://bugs.swift.org/browse/SR-13032

martinboehme commented:
@swift-ci Please test

gribozavr commented on include/swift/AST/PrintOptions.h:
```diff
@@ -548,7 +548,8 @@ struct PrintOptions {
   ///
   /// \see swift::emitSwiftInterface
-  static PrintOptions printSwiftInterfaceFile(bool preferTypeRepr,
+  static PrintOptions printSwiftInterfaceFile(ModuleDecl *M,
```
Please document the new parameter.

martinboehme replied:
I'm not sure; in this context, what does it mean for a module to be "current"? This makes sense only if the reader knows that `PrintOptions` contains a field with this name, but not everyone may know that.

In this context, the module being passed here is the module to be printed. So what do you think of either explaining this in the comment above or renaming `M` to `ModuleToPrint`?

gribozavr replied:
`ModuleToPrint` would be a good name.

martinboehme replied:
Done.

gribozavr commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -100,14 +100,16 @@ static bool contributesToParentTypeStorage(const AbstractStorageDecl *ASD) {
 }
 
-PrintOptions PrintOptions::printSwiftInterfaceFile(bool preferTypeRepr,
+PrintOptions PrintOptions::printSwiftInterfaceFile(ModuleDecl *M,
```
```suggestion
PrintOptions PrintOptions::printSwiftInterfaceFile(ModuleDecl *CurrentModule,
```

gribozavr commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -100,14 +100,16 @@ static bool contributesToParentTypeStorage(const AbstractStorageDecl *ASD) {
   result.PrintLongAttrsOnSeparateLines = true;
   result.TypeDefinitions = true;
   result.PrintIfConfig = false;
+  result.CurrentModule = M;
```
```suggestion
  result.CurrentModule = CurrentModule;
```

gribozavr commented on test/Interop/C/modules/repro.swift:
```diff
@@ -0,0 +1,5 @@
+// RUN: %empty-directory(%t)
```
Is there a better name than "repro" that we can use? Similarly, for other headers in this example. My best suggestion is to actually create a separate subdirectory just for this test since it requires lots of different input files, and their names don't make it obvious that they are related, that they create a particular test case interaction, and shouldn't be reused for other reasons.

martinboehme replied:
Makes sense. I've gone with the slightly shorter name `print-qualified-clang-types` -- WDYT?

martinboehme committed:
Give test a more suitable name and move it to a separate directory.

Putting the test in a separate directory makes it clearer that all of
the input files are related and create a particular test case
interaction.

gribozavr commented on test/Interop/C/modules/Inputs/SubModule.swift:
```diff
@@ -0,0 +1,2 @@
+import ForeignA
```
Could you rename it to something other than "SubModule"? "Submodule" has a particular meaning, but here it is not that.

martinboehme replied:
Donel

gribozavr commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -3586,19 +3589,81 @@ class TypePrinter : public TypeVisitor<TypePrinter> {
+    if (!VisibleClangModules) {
+      VisibleClangModules.emplace();
+      ModuleDecl::ImportFilter Filter = ModuleDecl::ImportFilterKind::Public;
+      Filter |= ModuleDecl::ImportFilterKind::Private;
```
I'm afraid the test case is not testing this subtle difference in filtering for modules imported directly vs. modules a couple hops away.

martinboehme replied:
I'm not sure I understand what you mean.

*  `MainModule` imports `HelperModule` privately, and this import should be considered
*  `HelperModule` imports `ForeignA` privately and `ForeignB` publicly, but only the latter should be considered.

So I would think that the test case does cover this. Can you clarify what else you think should be tested?

MForster replied:
@gribozavr, could you clarify what you meant?

gribozavr replied:
@martinboehme is right, the test covers both cases!

gribozavr commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -3586,19 +3589,81 @@ class TypePrinter : public TypeVisitor<TypePrinter> {
+    // Clang declarations need special treatment: Multiple Clang modules can
+    // contain the same declarations from a textually included header, but not
+    // all of these modules may be visible. We therefore need to make sure we
+    // choose a module that is visible from the current module. This is
+    // obviously only possible, however, if we know what the current module is.
```
```suggestion
    // choose a module that is visible from the current module. This is
    // possible to do only if we know what the current module is.
```

martinboehme replied:
Done.

Additionally, I've left out the "to do", as that seemed redundant.

gribozavr reviewed (approved):
LGTM, but I also would like @nkcsgexi to take a look.

martinboehme committed:
Rename SubModule to HelperModule.

martinboehme commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -3586,19 +3589,81 @@ class TypePrinter : public TypeVisitor<TypePrinter> {
+  /// through public (`@_exported`) imports.
+  /// `clangModules` is a map from each imported Clang module to the
+  /// corresponding Swift module.
+  static void getImportedClangModules(
```
> WDYT about `computeVisibleClangModules`? That would make a more obvious connection with `getVisibleClangModules` and emphasize the get vs. compute distinction.

I consciously made this distinction because this is a lower-level function that, on its own, doesn't really guarantee that you'll get exactly the visible modules; that only happens if you pass the right value for `filter`.

I think this concern may go away though once I rewrite the function as you suggest below.

> Could you change `getImportedClangModules` to return the may by value, and to keep a data recursion queue, and to keep a set of already-visited modules? Module interface printing can get called from performance-sensitive contexts in the IDE, so I'd rather not take chances with an exponential algorithm.

That makes sense. Not done yet, but I wanted to respond. I'll put a comment on this thread when I've made this change.

martinboehme replied:
> Could you change `getImportedClangModules` to return the may by value, and to keep a data recursion queue, and to keep a set of already-visited modules? Module interface printing can get called from performance-sensitive contexts in the IDE, so I'd rather not take chances with an exponential algorithm.

Done.

martinboehme committed:
Handle Clang submodules correctly.

martinboehme commented:
@swift-ci Please test

martinboehme committed:
Use an explicit stack instead of recursion to compute visible Clang modules.

Also, make sure we only process a module once.

martinboehme committed:
Various responses to review comments.

martinboehme committed:
Formatting fix.

martinboehme commented:
@swift-ci Please test

martinboehme commented on test/Interop/C/modules/repro.swift:
```diff
@@ -0,0 +1,5 @@
+// RUN: mkdir %t/sub_module %t/main_module
+// RUN: %target-swift-frontend -enable-library-evolution -swift-version 5 -emit-module -o %t/sub_module/SubModule.swiftmodule %S/Inputs/SubModule.swift -I %S/Inputs
+// RUN: %target-swift-frontend -enable-library-evolution -swift-version 5 -emit-module -o %t/main_module/MainModule.swiftmodule -emit-module-interface-path %t/main_module/MainModule.swiftinterface -I %t/sub_module %S/Inputs/MainModule.swift -I %S/Inputs
+// RUN: %target-swift-frontend -typecheck -swift-version 5 %t/main_module/MainModule.swiftinterface -I %t/sub_module -I %S/Inputs
```
Done. Good point about the FileCheck -- that makes this a lot more explicit.

martinboehme commented:
@gribozavr I've now addressed all of your review comments.

MForster commented:
@nkcsgexi, could you take a look at this PR?

xymus reviewed (approved):
LGTM! Sorry for the delay.

varungandhi-apple commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -3586,19 +3590,99 @@ class TypePrinter : public TypeVisitor<TypePrinter> {
+
+    llvm::DenseMap<const clang::Module *, ModuleDecl *> Result;
+
+    // For the current module, consider both private and public imports.
+    ModuleDecl::ImportFilter Filter = ModuleDecl::ImportFilterKind::Public;
+    Filter |= ModuleDecl::ImportFilterKind::Private;
+    SmallVector<ModuleDecl::ImportedModule, 4> Imports;
+    Options.CurrentModule->getImportedModules(Imports, Filter);
```
I think the `SPIAccessControl` and `ShadowedBySeparateOverlay` import kinds also need to be included here, as those Clang modules will be visible. cc @xymus @brentdax

brentdax replied:
`ShadowedBySeparateOverlay`, at least, should *not* be necessary; those modules will be visible transitively through the overlay's `@_exported import`.

xymus replied:
Right, this may change soon but for now `SPIAccessControl` imports should likely be included for the search from the current module.

hlopko replied:
We've added a test for SPIAccessControl, and we fixed ASTPrinter to visit @_spi modules. Thanks for the suggestion!

shahmishal commented:
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

varungandhi-apple commented:
@swift-ci Please test

hlopko committed:
Sync main

hlopko committed:
Add test for importing @_spi module

hlopko commented:
@swift-ci please test

hlopko commented:
I believe this PR is now good to go, all comments are addressed and we have 2 approvals already, so unless anybody comments, we will merge it tomorrow.

scentini committed:
Merge branch 'main' into module-issue-fix

scentini committed:
Enable test that failed without this PR

hlopko commented:
@swift-ci please test
